arctic/store/metadata_store.py

```diff
@@ -0,0 +1,186 @@
+from datetime import datetime as dt
+import logging
+
+import pandas as pd
+
+import bson
+import pymongo
+from ..decorators import mongo_retry
+from ..exceptions import NoDataFoundException
+
+from .bson_store import BSONStore
+
+logger = logging.getLogger(__name__)
+
+METADATA_STORE_TYPE = 'MetadataStore'
+
+class MetadataStore(BSONStore):
+    """
+    Metadata Store. This stores metadata with timestamps to allow temporal queries.
+
+    Entries are stored in the following format:
+        'symbol': symbol name
+        'metadata': metadata to be persisted
+        'start_time': when entry becomes effective
+        'end_time': (Optional) when entry expires. If not set, it is still in effect
+
+    For each symbol end_time of a entry should match start_time of the next one except for the current entry.
+    """
+
+    @classmethod
+    def initialize_library(cls, arctic_lib, hashed=True, **kwargs):
+        MetadataStore(arctic_lib)._ensure_index()
+        BSONStore.initialize_library(arctic_lib, hashed, **kwargs)
+
+    @mongo_retry
+    def _ensure_index(self):
+        self.create_index([('symbol', pymongo.ASCENDING), ('start_time', pymongo.DESCENDING)],
+                          unique=True, background=True)
+
+    def __init__(self, arctic_lib):
+        self._arctic_lib = arctic_lib
+        self._reset()
+
+    def _reset(self):
+        self._collection = self._arctic_lib.get_top_level_collection().metadata
+
+    def list_symbols(self):
+        return self.distinct('symbol')
+
+    def has_symbol(self, symbol):
+        return self.find_one({'symbol': symbol}) is not None
+
+    def read(self, symbol, history=False):
+        """
+        Return the metadata saved for a symbol
+
+        Parameters
+        ----------
+        symbol : `str`
+            symbol name for the item
+        history: `bool`
+            False: Returns current metadata (Default)
+            True: Returns all metadata entries
+
+        Returns
+        -------
+        metadata document if history=False
+        pandas.DateFrame containing timestamps and metadata entries if history=True
+        """
+        if not history:
+            res = self.find_one({'symbol': symbol}, sort=[('start_time', pymongo.DESCENDING)])
+            return res['metadata'] if res is not None else None
+        find = self.find({'symbol': symbol}, sort=[('start_time', pymongo.ASCENDING)])
+        times = []
+        entries = []
+        for item in find:
+            times.append(item['start_time'])
+            entries.append(item['metadata'])
+        return pd.DataFrame({'metadata': entries}, times)
+
+    def write_history(self, collection):
+        """
```

Comment:
add the ability to write history from a pd.DataFrame assuming same format as the output from read()